### DISCUSSION TITLE
An Ideation for Kubernetes-native Kafka Connect - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#300] [Announcements]
# An Ideation for Kubernetes-native Kafka Connect - Gunnar Morling

Kafka Connect, part of the Apache Kafka project, is a development framework and runtime for connectors which either ingest data into Kafka clusters (source connectors) or propagate data from Kafka into external systems (sink connectors). A diverse ecosystem of ready-made connectors has come to life on top of Kafka Connect, which lets you connect all kinds of data stores, APIs, and other systems to Kafka in a no-code approach.
 With the continued move towards running software in the cloud and on Kubernetes in particular, it’s just natural that many folks also try to run Kafka Connect on Kubernetes.

https://www.morling.dev/blog/ideation-kubernetes-native-kafka-connect/

**Comment by devinay** [comment]
Thanks for writing this down. This is something that me and my team have been discussing as well: How to containerize Kafka connect correctly. The aims are very similar to the ones here:
- Multi tenant support. Different users should be able to instantiate different connectors and not know about each other. They should have access to the logs of their connectors so that they can self-serve, scheduling should be fair and predictable
- Low management overhead: When these logical connectors are reconfigured, for example if tasks are bumped up or down, or if  batch size is increased or decreased, the response should be immediate with as few moving parts kicking in as possible.
- A corollary to the above: as few changes as possible, leverage existing code and paradigms without having to create new ones.
- Network isolation should be possible, so if a particular user can only access a database within a VPN, it should be possible to run the connector adjacent to it, or VPC peer their connector.
- Debuggability and observability: It should be possible to easily track and observe bad tasks and bad connectors and figure out why they're bad, specially when there's a large number of different kinds of connector run by different users. It should be possible to collect stats and logs on a per-task basis easily.
- Provable bin packing of pods to minimize costs

Interestingly, the final conclusion is the same : There needs to be one task per pod.
The way to achieve this is somewhat different:
- Have a single docker image per type of container, so for example Debezium source will be one docker image that would be referenced in k8s, similarly there would be one image for ElasticSearch etc.
- An instance of a connector, i.e the connector that starts up when a user asks for one, is a k8s deployment (not a statefulset) that can potentially autoscale. There's an operator that would spin up a k8s deployment when a request for a new connector comes in. Based on the kind of connector, it would reference the docker image in the k8s manifest. So if a user 'U' asked for a JDBC source connector, a deployment say 'U_JDBC_Source' would be spun up, initially with just one active pod in it. This pod is plain old distributed-connect with no code changes. 
- The resource limits in k8s are set based on the type of connector, for example a memory-heavy connector would have some more memory and a CPU heavy one would have more CPU. The important thing is that the resource sizing is on a one-task basis: How much memory CPU does this kind of connector need to run just one task ? There is some headroom for vertical scale. This is to acheive bin-packing.
- The connector would emit a tasks-per-pod metric, this would be used in HPA (horizontal pod autoscaling). The limit is set to 1

So when a user asks for an instance of a particular connector, the operator creates an autoscaleable deployment of connect-distributed with just one pod, referencing the dockerfile of that connector. Multiple tasks are spun up within one pod, but because of the autoscaling rule, multiple pods are started up within the deployment and each pod will have exactly one task in steady state. In case of pod failure, some tasks will migrate transparently to the existing pods but then autoscaling would kick in and steady state would be reached again.

Thoughts ?

**Reply by gunnarmorling** [reply]
Thanks a lot for this comment, that's super-interesting to read! In your model, are you running a separate cluster for each connector then? Or could it still happen that tasks of multiple connectors are co-located on one worker node (until the auto-scaler has spun up enough worker nodes)?

**Reply by devinay** [reply]
Each docker image in a k8s cluster's namespace is connector-specific, so it has a worker + <connector specific jar> only and runs just one task in the final state based on the tasks-per-worker metric. So (task/worker/pod) can be used interchangeably.

No, the plan is to co-locate disparate connectors with different resource requirements on the same k8s cluster, each in its own namespace, and let the auto-scaler kick in automatically based on tasks per worker.

Say for example there is a 'fat' connector A that uses a lot of resources and there's a 'thin' connector B that does not use as much. When user U1 asks for an instance of A with 5 tasks, a one-pod deployment is spun up for U1. Now since 5 tasks start and the average of the tasks-per-pod is 5, the autoscaler kicks in bringing up pods until the average drops to 1. If U2 asks for B with 3 tasks, the same thing happens on the same k8s cluster, so there's always one task per pod/worker in the final state, its not as if a worker never runs 2 tasks. That could happen intermittently, for example at startup, when the number of tasks is modified or if there's failure. But finally, there's always one task per pod. The difference between A and B is that A would be given more resources than B, but both run on the same cluster. The plan is to spin up a new cluster only when some external limits are hit, for example the I/O of all the connectors put together is too much or if there are some unforeseen k8s limits that start to come into play.
